### PR TITLE
[WIP] rfcs: Support for easier hash sharded indexes

### PR DESCRIPTION
**Pull request by aayushshah15**
This RFC proposes to provide better UX when creating hash sharded
indexes through easier syntax.

Release note: None

### CODE REVIEW

File: docs/RFCS/20191028_easier_hash_sharded_indexes.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+(2) Using primary index on `k`.
+```sql
+CREATE TABLE kv (
+    k INT8 NOT NULL,
+    v BYTES NOT NULL,
+    PRIMARY KEY (k)
+);
+```
+
+ The following benchmarks were all run on a 7 node GCE cluster with `n1-standard-32`
```
Comment (ajwerner) [inline docs/RFCS/20191028_easier_hash_sharded_indexes.md]: These are big machines, you should consider more shards. Play around with something like one shard per core or maybe a little bit less? Perhaps 64? Alternatively try smaller machines.

File: docs/RFCS/20191028_easier_hash_sharded_indexes.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts) USING HASH (product_id, owner, serial_number) WITH (BUCKET_COUNT=8)
+);
+```
+
+# Benchmarks
```
Comment (ajwerner) [inline docs/RFCS/20191028_easier_hash_sharded_indexes.md]: Other things to look at are performance for sharded secondary indexes which have sequential values and then also to look at range scan performance in both the primary and secondary index cases.

File: docs/RFCS/20191028_easier_hash_sharded_indexes.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    serial_number
+        VARCHAR,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts) USING HASH (product_id, owner, serial_number) WITH (BUCKET_COUNT=8)
```
Comment (ajwerner) [inline docs/RFCS/20191028_easier_hash_sharded_indexes.md]: This works for sharding a secondary index but make sure to add a proposal for how to shard a primary index. Furthermore do call out where this syntax is coming from by referencing postgres and SQL server docs as relevant.

File: docs/RFCS/20191028_easier_hash_sharded_indexes.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+ machines. We first demonstrate write performance with `kv --sequential --duration=5m` at
+ varying levels of concurrency (256, 384, 512). Note that this is a write-only workload.
+
+
+![Latency](https://user-images.githubusercontent.com/10788754/67722397-63c5a100-f9af-11e9-8316-a173e52fe20b.png)
+
+
+![Throughput](https://user-images.githubusercontent.com/10788754/67733735-8ec2eb80-f9d5-11e9-889f-38b4466d7234.png)
+
+## Drawbacks / Future work
```
Comment (ajwerner) [inline docs/RFCS/20191028_easier_hash_sharded_indexes.md]: We spoke offline but talk through what it takes to plumb the calculation of shard through to the select. 

The idea is that for these point lookups, when this project is actually done, we won't have this inefficiency.

File: docs/RFCS/20191028_easier_hash_sharded_indexes.md (hunk 1 above)
Comment (aayushshah15) [reply docs/RFCS/20191028_easier_hash_sharded_indexes.md]: okay, smaller machines seem like a better idea, I'll regenerate these numbers.

**Review by aayushshah15 — COMMENTED** [review]
I had a bit of a hard time passing custom benchmark output to `benchstat` (I tried to match the format of my files to match the format from `go test -bench`, but no dice). I'm adding simple line graphs for now. Currently trying to deflake some tests but I can create new RFC & prototype PRs (as we discussed) once I have your :+1: .

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-:-Lv7WZIk36jh7YgCd-Ja:b-fc82h0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajwerner)

___
*[docs/RFCS/20191028_easier_hash_sharded_indexes.md, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-LsJTYYRWHrvRCoT0i0s-r2-142:-Lv7XlrOCAkjSO1Mbl-n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc88c38081c0d17611fc40aabd920eba26e7f7e7/docs/RFCS/20191028_easier_hash_sharded_indexes.md#L142)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This works for sharding a secondary index but make sure to add a proposal for how to shard a primary index. Furthermore do call out where this syntax is coming from by referencing postgres and SQL server docs as relevant. 
</blockquote></details>

Done.
___
*[docs/RFCS/20191028_easier_hash_sharded_indexes.md, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-LsJTYYRWHrvRCoT0i0s-r2-181:-Lv7ZsqJ3OCQIniLmCJn:b9zcpmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc88c38081c0d17611fc40aabd920eba26e7f7e7/docs/RFCS/20191028_easier_hash_sharded_indexes.md#L181)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

We spoke offline but talk through what it takes to plumb the calculation of shard through to the select. 

The idea is that for these point lookups, when this project is actually done, we won't have this inefficiency. 
</blockquote></details>

Added this, let me know if it's not what you meant.

**Review by ajwerner — COMMENTED** [review]
> I had a bit of a hard time passing custom benchmark output to benchstat (I tried to match the format of my files to match the format from go test -bench, but no dice). I'm adding simple line graphs for now.

ack. A bar chart might do you better than the line chart for this data.

>  Currently trying to deflake some tests but I can create new RFC & prototype PRs (as we discussed) once I have your 👍 .

I have some nits but they can wait for the real PR. I think it's time to split this into a real RFC PR and an implementation PR. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-:-Lv7xFR972NcdCGre291:b-8rl8dg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajwerner)

___
*[docs/RFCS/20191028_easier_hash_sharded_indexes.md, line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-Lv7yIKZ3IYsB7QhCBz3:-Lv7yIKZ3IYsB7QhCBz4:b-smx5xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ea92e848d581c04616c3727307f6a7bc78265/docs/RFCS/20191028_easier_hash_sharded_indexes.md#L149)):*
> ```Markdown
>     -- Creates a primary index on (shard, product_id, owner, serial_number)
>     -- with a check constraint for `shard in (0...7)`
>     PRIMARY KEY (product_id, owner, serial_number) USING HASH WITH BUCKET_COUNT=8
> ```

We have https://www.cockroachlabs.com/docs/stable/sql-grammar.html#opt_with_options in the grammar already. Seems like it would be better to follow that convention than adding the BUCKET_COUNT as an actual token.
___
*[docs/RFCS/20191028_easier_hash_sharded_indexes.md, line 195 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-Lv7y6Q2AEsl8E2Qu_Gp:-Lv7y6Q2AEsl8E2Qu_Gq:bpir4n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ea92e848d581c04616c3727307f6a7bc78265/docs/RFCS/20191028_easier_hash_sharded_indexes.md#L195)):*
> ```Markdown
> ## Benchmarks 
> 
>  The following benchmarks were all run on a 5 node GCE cluster with `n1-standard-8`
> ```

If we really want to make this proposal impressive we could add another series with a much larger cluster to demonstrate that this proposal unlocks horizontal scalability.

**Review by aayushshah15 — COMMENTED** [review]
I just separated this into two PRs. Will move the rest of your comments on to the respective PRs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-:-LvCXMxj2Vo7MSQ5mg6c:b-ut3w9a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajwerner)

___
*[docs/RFCS/20191028_easier_hash_sharded_indexes.md, line 195 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41974#-Lv7y6Q2AEsl8E2Qu_Gp:-LvCXJ6u8tw_3BNsDRJA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ea92e848d581c04616c3727307f6a7bc78265/docs/RFCS/20191028_easier_hash_sharded_indexes.md#L195)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

If we really want to make this proposal impressive we could add another series with a much larger cluster to demonstrate that this proposal unlocks horizontal scalability. 
</blockquote></details>

Done.

**Pull request closed without merging**